subsys/bluetooth/audio/vocs_client.c

```diff
@@ -0,0 +1,688 @@
+			BT_DBG("Offset %d, counter %u",
+			       inst->cli.state.offset,
+			       inst->cli.state.change_counter);
+
+			/* clear busy flag to reuse function */
+			inst->cli.busy = false;
+			write_err = bt_vocs_client_state_set(conn, inst, inst->cli.cp.offset);
+			if (write_err) {
+				cb_err = BT_ATT_ERR_UNLIKELY;
+			}
+		} else {
+			BT_DBG("Invalid length %u (expected %zu)", length, sizeof(inst->cli.state));
+			cb_err = BT_ATT_ERR_UNLIKELY;
+		}
+	} else {
+		BT_DBG("Invalid location");
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	}
+
+	if (cb_err) {
+		inst->cli.busy = false;
+
+		if (inst->cli.cb && inst->cli.cb->set_offset) {
+			inst->cli.cb->set_offset(conn, inst, err);
+		}
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static void vcs_client_write_vocs_cp_cb(struct bt_conn *conn, uint8_t err,
+					struct bt_gatt_write_params *params)
+{
+	int cb_err = err;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, params->handle);
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return;
+	}
+
+	BT_DBG("Inst %p: err: 0x%02X", inst, err);
+
+	/* If the change counter is out of data when a write was attempted from the application,
+	 * we automatically initiate a read to get the newest state and try again. Once the
+	 * change counter has been read, we restart the applications write request. If it fails
+	 * the second time, we return an error to the application.
+	 */
+	if (cb_err == BT_VOCS_ERR_INVALID_COUNTER && inst->cli.state_handle) {
+		BT_DBG("Invalid change counter. Reading volume state from server.");
+
+		inst->cli.read_params.func = internal_read_volume_offset_state_cb;
+		inst->cli.read_params.handle_count = 1;
+		inst->cli.read_params.single.handle = inst->cli.state_handle;
+
+		cb_err = bt_gatt_read(conn, &inst->cli.read_params);
+
+		if (cb_err) {
+			BT_WARN("Could not read Volume state: %d", cb_err);
+		}
+	} else {
+		BT_DBG("Invalid location");
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	}
+
+	inst->cli.busy = false;
+
+	if (inst->cli.cb && inst->cli.cb->set_offset) {
+		inst->cli.cb->set_offset(conn, inst, cb_err);
+	}
+}
+
+static uint8_t vcs_client_read_output_desc_cb(struct bt_conn *conn, uint8_t err,
+					      struct bt_gatt_read_params *params,
+					      const void *data, uint16_t length)
+{
+	int cb_err = err;
+	struct bt_vocs *inst = lookup_vocs_by_handle(conn, params->single.handle);
+	char desc[MIN(CONFIG_BT_L2CAP_RX_MTU, BT_ATT_MAX_ATTRIBUTE_LEN) + 1];
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("Inst %p: err: 0x%02X", inst, err);
+	inst->cli.busy = false;
+
+	if (data) {
+		BT_HEXDUMP_DBG(data, length, "Output description read");
+		length = MIN(sizeof(desc) - 1, length);
+
+		/* TODO: Handle long reads */
+		memcpy(desc, data, length);
+		desc[length] = '\0';
+		BT_DBG("Output description: %s", log_strdup(desc));
+	} else {
+		BT_DBG("Invalid location");
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	}
```

Comment:
What does `data == NULL` mean here? What is the stack trying to tell us?

Reply:
That there is no data; i.e. the GATT server sent invalid data on our read. This is application specific (some characteristics may be of length 0, so we cannot do anything about this below the application level).